runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -127,6 +125,49 @@ public ILogData peek(RuntimeLayout runtimeLayout, long globalAddress) {
         return waitOrHoleFill(runtimeLayout, readResult, waitForWrite);
     }
 
+    /**
+     * Commit the addresses by first reading and then hole filling if data not existed.
+     *
+     * @param runtimeLayout the RuntimeLayout stamped with layout to use for commit
+     * @param addresses     a collection of addresses to commit
+     */
+    @Override
+    public void commitAll(RuntimeLayout runtimeLayout, Collection<Long> addresses) {
+        // Group addresses by log unit client.
+        Map<LogUnitClient, List<Long>> serverAddressMap =
+                groupAddressByLogUnit(runtimeLayout, addresses);
+
+        // Send inspect addresses requests to log unit servers in parallel.
+        List<CompletableFuture<InspectAddressesResponse>> futures = serverAddressMap
+                .entrySet()
+                .stream()
+                .map(entry -> entry.getKey().inspectAddresses(entry.getValue()))
+                .collect(Collectors.toList());
+
+        // Merge the inspect responses from different log unit servers.
+        List<Long> holes = futures.stream()
+                .flatMap(future -> CFUtils.getUninterruptibly(future).getEmptyAddresses().stream())
+                .collect(Collectors.toList());
```

Comment:
How it handles TrimException in inspectAddresses?

Reply:
TrimmedException is poped up tp AddressSpaceView and then AutoCommitService. Then current commit cycle fails and retries. There's actually a bug in handling retry, I fixed will add a test case.